components/timezone/src/types.rs

```diff
@@ -35,6 +48,94 @@ impl GmtOffset {
+    /// let offset1: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"+05").expect("Failed to parse a time zone");
+    /// let offset2: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"+0500").expect("Failed to parse a time zone");
+    /// let offset3: GmtOffset =
+    ///     GmtOffset::try_from_bytes(b"-05:00").expect("Failed to parse a time zone");
+    /// let offset_err0: TimeZoneError =
+    ///     GmtOffset::try_from_bytes(b"0500").expect_err("Invalid input");
+    /// let offset_err1: TimeZoneError =
+    ///     GmtOffset::try_from_bytes(b"+05000").expect_err("Invalid input");
+    ///
+    /// assert_eq!(offset0.offset_seconds(), 0);
+    /// assert_eq!(offset1.offset_seconds(), 18000);
+    /// assert_eq!(offset2.offset_seconds(), 18000);
+    /// assert_eq!(offset3.offset_seconds(), -18000);
+    /// assert_eq!(offset_err0, TimeZoneError::InvalidOffset);
+    /// assert_eq!(offset_err1, TimeZoneError::InvalidOffset);
+    /// ```
+    pub fn try_from_bytes(chars: &[u8]) -> Result<Self, TimeZoneError> {
+        let mut it = chars.iter();
+        let offset_sign = match it.next() {
+            Some(b'+') => 1,
+            Some(b'-') => -1,
+            Some(b'Z') => return Ok(Self(0)),
+            _ => return Err(TimeZoneError::InvalidOffset),
+        };
+
+        let chars = it.as_slice();
+        let seconds = match chars.len() {
+            /* ±hh */
+            2 => {
+                let hour = try_get_time_component(chars).ok_or(TimeZoneError::InvalidOffset)?;
+                if hour > 24 {
+                    return Err(TimeZoneError::InvalidOffset);
+                }
+                offset_sign * (hour * 60 * 60)
+            }
+            /* ±hhmm */
+            4 => {
+                let (hour_slice, minute_slice) = chars.split_at(2);
+                let hour =
+                    try_get_time_component(hour_slice).ok_or(TimeZoneError::InvalidOffset)?;
+                let minute =
+                    try_get_time_component(minute_slice).ok_or(TimeZoneError::InvalidOffset)?;
+                offset_sign * (hour * 60 * 60 + minute * 60)
+            }
+            /* ±hh:mm */
+            5 => {
+                let mut splits = chars.split(|c| *c == b':');
+                let hour = splits
+                    .next()
+                    .and_then(self::try_get_time_component)
+                    .ok_or(TimeZoneError::InvalidOffset)?;
+                if hour > 24 {
+                    return Err(TimeZoneError::InvalidOffset);
+                }
+                let minute = splits
+                    .next()
+                    .and_then(self::try_get_time_component)
+                    .ok_or(TimeZoneError::InvalidOffset)?;
+                if minute > 60 {
+                    return Err(TimeZoneError::InvalidOffset);
+                }
+                offset_sign * (hour * 60 * 60 + minute * 60)
+            }
+            _ => return Err(TimeZoneError::InvalidOffset),
+        };
```

Comment:
Match on `chars` instead of `chars.len()`. Then you can make `try_get_time_component` accept `[u8; 2]` (or `u8`, `u8`) so you don't have to verify a length in there.

```rust
let seconds = match chars { 
  /* ±hh */
  &[h1, h2] => {
        let hour = try_get_time_component([h1, h2]).ok_or(TimeZoneError::InvalidOffset)?;
        if hour > 24 {
            return Err(TimeZoneError::InvalidOffset);
        }
        offset_sign * (hour * 60 * 60)
    }
    /* ±hhmm, ±hh:mm */
    &[h1, h2, m1, m2] | &[h1, h2, b':', m1, m2] => {
        let hour =
            try_get_time_component([h1, h2]).ok_or(TimeZoneError::InvalidOffset)?;
        if hour > 24 {
            return Err(TimeZoneError::InvalidOffset);
        }
        let minute =
            try_get_time_component([m1, m2]).ok_or(TimeZoneError::InvalidOffset)?;
        if minute > 60 {
            return Err(TimeZoneError::InvalidOffset);
        }
        offset_sign * (hour * 60 * 60 + minute * 60)
    }
    _ => return Err(TimeZoneError::InvalidOffset),
};
```

Reply:
Love slice pattern matching!